_posts/2020-18-06-Triplestores-Onotologies-Reasoning.md

```diff
@@ -0,0 +1,354 @@
+predicate, a Triplestore that implements OWL 2 will _infer_ that the
+URL belongs to a particular class without further intervention from a
+user.  Thus, using the turtle abbreviation `a` is often not necessary
+as the Triplestore will infer the type based on its use in the RDF.
+This kind of inference is often called _reasoning_ and a system that
+implements the logical system of OWL 2 is called a _reasoner`_.  There
+are several reasoners that are independent of a Triplestore and can be
+used without a storage system such as
+[Fact++](https://doi.org/10.1007/11814771_26),
+[HermiT](http://www.hermit-reasoner.com/), and Pellet (which happens
+to be the reasoner used by Stardog).
+
+```turtle
+<#h-ÚiBuide> irishRel:populationGroupName "h-Úi Buide".
+```
+
+Looking again at the above triple, the effect of the OWL 2 definition
+is that a Triplestore's reasoner will infer that `<#h-ÚiBuide>` is a
+`PopulationGroup` and infer that `"h-Úi Buide"` is a string.  A note
+of caution is appropriate at this point.  While a Triplestore with a
+reasoner will infer, it will _not_ enforce.  Thus, for instance if a
+user puts a number literal, like _42_, into the object position in the
+above triple, a reasoner will correctly ingest this and not inform the
+user.  Enforcement or validation of triples can be done using the
+[Shapes Constraint Language (SHACL)](https://www.w3.org/TR/shacl/)
+specification if the Triplestore supports it.  Additionally, if two
+predicates define different classes for their respective URLs, a user
+can end up in a position where a reasoner infers a `Person` class and
+a `Cat` class on the same URL, for instance, which will confuse a user
+who expecting a `Person` and a `Cat` to be distinct individuals.  The
+curators need to be careful when using predicates to ensure they do
+not cause odd or confusing inferences to be made.
+
+However, there are even more powerful features of OWL 2.  Predicates
+that can be defined as
+[_transitive_](https://www.w3.org/TR/2012/REC-owl2-primer-20121211/#Property_Characteristics)
+in OWL which means that if a predicate connects A and B while B uses
+the same predicate to connect B to C then A has the same effect as if
+A also connected to C.
+
+An instance from IrishGen should make this more clear.
+
+```turtle
+<#Laidcnén>
+    a foaf:Person;
+    irishRel:nomName "Laidcnén";
+    rel:childOf <#Cummine>;
+    irishRel:ancestorOfGroup <#h-ÚiBuide>.
+```
+
+IrishGen heavily uses the Relationship ontology which defines
+`rel:parentOf` and `rel:childOf`.  These are defined as a
+`rdf:subPropertyOf` `rel:ancestorOf` and `rel:descendantOf`.
+`rel:ancestor` and `rel:descendantOf` are themselves defined as
+transitive and the inverse of each other.  Summing all of that up, it
+means that a reasoner can completely reconstruct a genealogical line
+without human intervention.  An instance of this will be illustrative.
+Without needing to encode by hand Laidcnén has 2301 ancestors.
```

Comment:
"Without needing to encode by hand Laidcnén has 2301 ancestors" > "Without needing to encode the information by hand, it can be determined via a query Laidcnén has 2301 ancestors" or suchlike?